File: src/notification_tray/components/notification_cacher.py
import json
import logging
from datetime import UTC, datetime
from pathlib import Path
from threading import Thread

from PyQt5.QtCore import QObject, pyqtSignal  # type: ignore
from send2trash import send2trash as _send2trash

from notification_tray.components.notifier import Notifier
from notification_tray.types.notification import (CachedNotification,
                                                  NotificationFolder)
from notification_tray.utils import settings
from notification_tray.utils.logging import log_input_and_output

send2trash = log_input_and_output(logging.INFO)(_send2trash)

logger = logging.getLogger(__name__)


class NotificationCacher(QObject):
    notifications_cached = pyqtSignal()

    def __init__(
        self,
        root_path: Path,
        notifier: Notifier,
        do_not_disturb: settings.Cache,
        notification_backoff_minutes: dict[Path, int],
        notification_cache: NotificationFolder,
        parent: QObject | None = None,
    ):
        super().__init__(parent)
        self.root_path = root_path
        self.notifier = notifier
        self.notification_cache = notification_cache
        self.do_not_disturb = do_not_disturb
        self.notification_backoff_minutes = notification_backoff_minutes
        logger.info(f"Started notification cacher with root path {root_path}")

    def cache_existing_notifications(self, root_path: Path):
        logger.info(f"Caching existing notifications under {root_path}")
        for dirpath, _, filenames in root_path.walk():
            dirpath = Path(dirpath)
            filenames = [
                f for f in filenames if f.endswith(".json") and f != ".settings.json"
            ]
            if filenames:
                logger.info(f"Found {len(filenames)} notifications under {dirpath}")
                notifications = self.notification_cache
                for folder in dirpath.relative_to(root_path).parts:
                    notifications = notifications["folders"].setdefault(
                        folder,
                        NotificationFolder(
                            folders={},
                            notifications={},
                            path=notifications["path"] / folder,
                        ),
                    )
                for filename in filenames:
                    path = dirpath / filename
                    try:
                        logger.debug(f"Loading {path}")
                        notifications["notifications"][filename] = CachedNotification(
                            json.loads(path.read_text())
                            | {
                                "path": path,
                                "at": datetime.fromtimestamp(path.stat().st_mtime, UTC),
                            }
                        )
                    except Exception as e:
                        logging.exception(f"Got exception at file {path}: {e}")
                        raise
        self.notifications_cached.emit()

    def cache(self, notification: CachedNotification):
        notifications = self.notification_cache
        if not notification["hints"].get("transient"):
            notification["path"].parent.mkdir(parents=True, exist_ok=True)
            notification_to_dump = dict(notification.copy())
            del notification_to_dump["at"]
            del notification_to_dump["path"]
            with notification["path"].open("w") as f:
                json.dump(notification_to_dump, f)
            logging.info(
                f"Notification {notification['summary']} written to {notification['path']}",
            )

        for folder in (
            notification["path"]
            .relative_to(self.notification_cache["path"])
            .parent.parts
        ):
            notifications = notifications["folders"].setdefault(
                folder,
                NotificationFolder(
                    folders={}, notifications={}, path=notifications["path"] / folder
                ),
            )
        path = notification["path"]
        notifications["notifications"][path.name] = notification
        self.notifications_cached.emit()

    @log_input_and_output(logging.INFO)
    def trash(self, path: Path):
        def mark_as_trashed(folder: NotificationFolder):
            for notification in folder["notifications"].values():
                notification["trashed"] = True
            for subfolder in folder["folders"].values():
                mark_as_trashed(subfolder)

        if path.exists():
            notifications = self.notification_cache
            for folder in path.relative_to(self.root_path).parent.parts:
                notifications = notifications["folders"][folder]
            if path.is_file():
                if path.suffix == ".json" and path.name != ".settings.json":
                    send2trash(path)
                    notifications["notifications"][path.name]["trashed"] = True
            else:
                if not list(path.rglob(".settings.json")) and not list(
                    path.rglob(".notification.wav")
                ):
                    send2trash(path)
                    mark_as_trashed(notifications["folders"][path.name])
                else:
                    if path != self.root_path:
                        notifications = notifications["folders"][path.name]
                    for folder in notifications["folders"].values():
                        Thread(
                            target=self.trash, args=(folder["path"],), daemon=True
                        ).start()
                    for file in notifications["notifications"]:
                        Thread(
                            target=self.trash, args=(path / file,), daemon=True
                        ).start()

            self.notifications_cached.emit()
        else:
            logger.info(f"Path does not exist. Skipping {path}")

File: src/notification_cacher.cpp
#include "notification_cacher.h"

#include "notifier.h"
#include "utils/logging.h"

#include <QDir>
#include <QFile>
#include <QFileInfo>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>

#include <thread>

// Improved send2trash implementation using Qt
static bool send2trash(const fs::path& path) {
    QString qpath = QString::fromStdString(path.string());

#if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
    // Qt 5.15+ has built-in moveToTrash
    return QFile::moveToTrash(qpath);
#else
    // Fallback: Move to freedesktop.org trash location
    QString trash_dir = QDir::homePath() + "/.local/share/Trash/files";
    QString info_dir = QDir::homePath() + "/.local/share/Trash/info";

    QDir().mkpath(trash_dir);
    QDir().mkpath(info_dir);

    QFileInfo file_info(qpath);
    QString filename = file_info.fileName();
    QString trash_path = trash_dir + "/" + filename;
    QString info_path = info_dir + "/" + filename + ".trashinfo";

    // Handle name conflicts
    int counter = 1;
    while (QFile::exists(trash_path)) {
        trash_path = trash_dir + "/" + filename + "." + QString::number(counter);
        info_path = info_dir + "/" + filename + "." + QString::number(counter) + ".trashinfo";
        counter++;
    }

    // Create .trashinfo file (freedesktop.org trash specification)
    QFile info_file(info_path);
    if (info_file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        QTextStream out(&info_file);
        out << "[Trash Info]\n";
        out << "Path=" << file_info.absoluteFilePath() << "\n";
        out << "DeletionDate=" << QDateTime::currentDateTime().toString(Qt::ISODate) << "\n";
        info_file.close();
    }

    // Move file to trash
    if (QFile::exists(qpath)) {
        if (file_info.isDir()) {
            return QDir().rename(qpath, trash_path);
        } else {
            return QFile::rename(qpath, trash_path);
        }
    }

    return false;
#endif
}

static Logger logger = Logger::getLogger("NotificationCacher");

NotificationCacher::NotificationCacher(const fs::path& root_path, Notifier* notifier,
                                       const Cache& do_not_disturb,
                                       const std::map<fs::path, int>& notification_backoff_minutes,
                                       NotificationFolder& notification_cache,
                                       const QString& run_id, QObject* parent)
    : QObject(parent)
    , m_root_path(root_path)
    , m_notifier(notifier)
    , notification_cache(notification_cache)
    , m_do_not_disturb(do_not_disturb)
    , m_notification_backoff_minutes(notification_backoff_minutes)
    , m_run_id(run_id) {
    logger.info(QString("Started notification cacher with root path %1")
                    .arg(QString::fromStdString(root_path.string())));
}

void NotificationCacher::cacheExistingNotifications(const fs::path& root_path) {
    logger.info(QString("Caching existing notifications under %1")
                    .arg(QString::fromStdString(root_path.string())));

    for (const auto& entry : fs::recursive_directory_iterator(root_path)) {
        if (entry.is_regular_file() && entry.path().extension() == ".json" &&
            entry.path().filename() != ".settings.json") {
            fs::path dirpath = entry.path().parent_path();
            fs::path relative = fs::relative(dirpath, root_path);

            NotificationFolder* current = &notification_cache;
            for (const auto& part : relative) {
                QString folder_name = QString::fromStdString(part.string());
                if (!current->folders.count(folder_name)) {
                    current->folders[folder_name].path = current->path / part;
                }
                current = &current->folders[folder_name];
            }

            QFile file(QString::fromStdString(entry.path().string()));
            if (file.open(QIODevice::ReadOnly)) {
                QJsonDocument doc = QJsonDocument::fromJson(file.readAll());
                if (doc.isObject()) {
                    QJsonObject obj = doc.object();

                    CachedNotification notif;
                    notif.app_name = obj["app_name"].toString();
                    notif.summary = obj["summary"].toString();
                    notif.body = obj["body"].toString();
                    notif.app_icon = obj["app_icon"].toString();
                    notif.id = obj["id"].toInt();
                    notif.replaces_id = obj["replaces_id"].toInt();
                    notif.expire_timeout = obj["expire_timeout"].toInt();
                    notif.notification_tray_run_id = obj["notification_tray_run_id"].toString();
                    notif.path = entry.path();

                    // Load actions from JSON
                    if (obj.contains("actions") && obj["actions"].isObject()) {
                        QJsonObject actions_obj = obj["actions"].toObject();
                        for (auto it = actions_obj.begin(); it != actions_obj.end(); ++it) {
                            notif.actions[it.key()] = it.value().toString();
                        }
                    }

                    // Load hints from JSON
                    if (obj.contains("hints") && obj["hints"].isObject()) {
                        QJsonObject hints_obj = obj["hints"].toObject();
                        for (auto it = hints_obj.begin(); it != hints_obj.end(); ++it) {
                            QJsonValue val = it.value();
                            if (val.isString()) {
                                notif.hints[it.key()] = val.toString();
                            } else if (val.isDouble()) {
                                // JSON doesn't distinguish int/double, check if it's a whole number
                                double d = val.toDouble();
                                if (d == static_cast<int>(d)) {
                                    notif.hints[it.key()] = static_cast<int>(d);
                                } else {
                                    notif.hints[it.key()] = d;
                                }
                            } else if (val.isBool()) {
                                notif.hints[it.key()] = val.toBool();
                            }
                        }
                    }

                    QFileInfo fileInfo(file);
                    notif.at = fileInfo.lastModified().toUTC();

                    current
                        ->notifications[QString::fromStdString(entry.path().filename().string())] =
                        notif;
                }
            }
        }
    }

    emit notificationsCached();
}

void NotificationCacher::cache(const CachedNotification& notification) {
    if (!notification.hints.value("transient", false).toBool()) {
        fs::create_directories(notification.path.parent_path());

        QJsonObject obj;
        obj["app_name"] = notification.app_name;
        obj["summary"] = notification.summary;
        obj["body"] = notification.body;
        obj["app_icon"] = notification.app_icon;
        obj["id"] = notification.id;
        obj["replaces_id"] = notification.replaces_id;
        obj["expire_timeout"] = notification.expire_timeout;
        obj["notification_tray_run_id"] = notification.notification_tray_run_id;

        // Serialize actions to JSON object
        QJsonObject actions_obj;
        for (auto it = notification.actions.begin(); it != notification.actions.end(); ++it) {
            actions_obj[it->first] = it->second;
        }
        obj["actions"] = actions_obj;

        // Serialize hints to JSON object
        QJsonObject hints_obj;
        for (auto it = notification.hints.begin(); it != notification.hints.end(); ++it) {
            QVariant value = it.value();
            // Convert QVariant to appropriate JSON type
            if (value.type() == QVariant::String) {
                hints_obj[it.key()] = value.toString();
            } else if (value.type() == QVariant::Int || value.type() == QVariant::LongLong) {
                hints_obj[it.key()] = value.toLongLong();
            } else if (value.type() == QVariant::UInt || value.type() == QVariant::ULongLong) {
                hints_obj[it.key()] = static_cast<qint64>(value.toULongLong());
            } else if (value.type() == QVariant::Double) {
                hints_obj[it.key()] = value.toDouble();
            } else if (value.type() == QVariant::Bool) {
                hints_obj[it.key()] = value.toBool();
            } else if (value.canConvert<QString>()) {
                hints_obj[it.key()] = value.toString();
            }
            // Skip complex types like image-data that can't be easily serialized
        }
        obj["hints"] = hints_obj;

        QFile file(QString::fromStdString(notification.path.string()));
        if (file.open(QIODevice::WriteOnly)) {
            file.write(QJsonDocument(obj).toJson());
            logger.info(
                QString("Notification %1 written to %2")
                    .arg(notification.summary, QString::fromStdString(notification.path.string())));
        }
    }

    fs::path relative = fs::relative(notification.path.parent_path(), notification_cache.path);
    NotificationFolder* current = &notification_cache;
    for (const auto& part : relative) {
        QString folder_name = QString::fromStdString(part.string());
        if (!current->folders.count(folder_name)) {
            current->folders[folder_name].path = current->path / part;
        }
        current = &current->folders[folder_name];
    }

    current->notifications[QString::fromStdString(notification.path.filename().string())] =
        notification;
    emit notificationsCached();
}

void NotificationCacher::trash(const fs::path& path) {
    logger.info(QString("Trashing %1").arg(QString::fromStdString(path.string())));

    if (path == m_root_path) {
        for (auto& [name, subfolder] : notification_cache.folders) {
            std::thread([this, subfolder]() { trash(subfolder.path); }).detach();
        }
        for (auto& [name, notif] : notification_cache.notifications) {
            std::thread([this, notif]() { trash(notif.path); }).detach();
        }
        emit notificationsCached();
        return;
    }

    fs::path relative = fs::relative(path.parent_path(), m_root_path);
    NotificationFolder* current = &notification_cache;
    for (const auto& part : relative) {
        QString folder_name = QString::fromStdString(part.string());
        if (current->folders.count(folder_name)) {
            current = &current->folders[folder_name];
        }
    }

    if (fs::is_regular_file(path)) {
        if (path.extension() == ".json" && path.filename() != ".settings.json") {
            send2trash(path);
            QString filename = QString::fromStdString(path.filename().string());
            if (current->notifications.count(filename)) {
                current->notifications[filename].trashed = true;
                if (current->notifications[filename].notification_tray_run_id == m_run_id) {
                    emit notificationTrashed(current->notifications[filename].id);
                }
            }
        }
    } else if (!fs::exists(path)) {
        // Handle transient notifications that exist in cache but not on disk
        QString filename = QString::fromStdString(path.filename().string());
        if (current->notifications.count(filename)) {
            logger.info(QString("Marking transient notification %1 as trashed").arg(filename));
            current->notifications[filename].trashed = true;
            if (current->notifications[filename].notification_tray_run_id == m_run_id) {
                emit notificationTrashed(current->notifications[filename].id);
            }
        }
    } else {
        bool has_settings = false;
        for (const auto& entry : fs::recursive_directory_iterator(path)) {
            if (entry.path().filename() == ".settings.json" ||
                entry.path().filename() == ".notification.wav") {
                has_settings = true;
                break;
            }
        }

        if (!has_settings) {
            send2trash(path);
            QString folder_name = QString::fromStdString(path.filename().string());
            if (current->folders.count(folder_name)) {
                markAsTrashed(current->folders[folder_name]);
            }
        } else {
            QString folder_name = QString::fromStdString(path.filename().string());
            if (current->folders.count(folder_name)) {
                NotificationFolder& folder = current->folders[folder_name];
                for (auto& [name, subfolder] : folder.folders) {
                    std::thread([this, subfolder]() { trash(subfolder.path); }).detach();
                }
                for (auto& [name, notif] : folder.notifications) {
                    std::thread([this, notif]() { trash(notif.path); }).detach();
                }
            }
        }
    }

    emit notificationsCached();
}

void NotificationCacher::markAsTrashed(NotificationFolder& folder) {
    for (auto& [name, notif] : folder.notifications) {
        notif.trashed = true;
        if (notif.notification_tray_run_id == m_run_id) {
            emit notificationTrashed(notif.id);
        }
    }
    for (auto& [name, subfolder] : folder.folders) {
        markAsTrashed(subfolder);
    }
}
